cmd/kubectl-directpv/migrate.go

```diff
@@ -0,0 +1,382 @@
+		for {
+			result, err := directCSIVolumeClient.List(ctx, options)
+			if err != nil {
+				send(ListVolumeResult{Err: err})
+				return
+			}
+
+			for _, item := range result.Items {
+				if !send(ListVolumeResult{Volume: item}) {
+					return
+				}
+			}
+
+			if result.Continue == "" {
+				return
+			}
+
+			options.Continue = result.Continue
+		}
+	}()
+
+	return resultCh
+}
+
+func validateMigrateCmd() error {
+	return nil
+}
+
+func migrateVolumeCRD(oldVolumeList []directcsi.DirectCSIVolume) (newVolumeList []directpv.DirectPVVolume, theerror error) {
+	newVolumes := []directpv.DirectPVVolume{}
+	for _, OldVolume := range oldVolumeList {
+		// FIXME: use types.NewDrive() instead of creating structure.
+		NewVolume := directpv.DirectPVVolume{}
+
+		// TypeMeta fields
+		NewVolume.TypeMeta.Kind = VolumeTypeMetaKind
+		NewVolume.TypeMeta.APIVersion = APIVersion
+
+		// ObjectMeta fields
+		NewVolume.ObjectMeta = OldVolume.ObjectMeta
+
+		// Status fields
+		NewVolume.Status.DataPath = OldVolume.Status.HostPath
+		NewVolume.Status.StagingTargetPath = OldVolume.Status.StagingPath
+		NewVolume.Status.TargetPath = OldVolume.Status.ContainerPath
+		NewVolume.Status.FSUUID = driveFsuuid[OldVolume.Status.Drive]
+		NewVolume.Status.TotalCapacity = OldVolume.Status.TotalCapacity
+		NewVolume.Status.AvailableCapacity = OldVolume.Status.AvailableCapacity
+		NewVolume.Status.UsedCapacity = OldVolume.Status.UsedCapacity
+
+		// Append the new drive to the list
+		newVolumes = append(newVolumes, NewVolume)
+	}
+	return newVolumes, nil
+}
+
+func migrateDriveCRD(oldDriveList []directcsi.DirectCSIDrive) (newDriveList []directpv.DirectPVDrive, therror error) {
+	// Set of unique FSUUIDs:
+	setOfUniqueFSUUIDs := make(map[string]struct{})
+
+	newDrives := []directpv.DirectPVDrive{}
+	for _, OldDrive := range oldDriveList {
+
+		// we can filter out the drives other than Ready and InUse
+		if OldDrive.Status.DriveStatus != directcsi.DriveStatusReady && OldDrive.Status.DriveStatus != directcsi.DriveStatusInUse {
+			continue // skip this iteration to filter out this state.
+		}
+
+		// FIXME: use types.NewVolume() instead of creating structure.
+		NewDrive := directpv.DirectPVDrive{}
+
+		// TypeMeta fields
+		NewDrive.TypeMeta.Kind = DriveTypeMetaKind
+		NewDrive.TypeMeta.APIVersion = APIVersion
+
+		// ObjectMeta fields
+		NewDrive.ObjectMeta = OldDrive.ObjectMeta
+
+		// Status fields
+		// Cast from old type to new type
+		NewDrive.Status.Status = types.DriveStatusReady
+		NewDrive.Status.TotalCapacity = OldDrive.Status.TotalCapacity
+		NewDrive.Status.AllocatedCapacity = OldDrive.Status.AllocatedCapacity
+		NewDrive.Status.FreeCapacity = OldDrive.Status.FreeCapacity
+		NewDrive.Status.FSUUID = OldDrive.Status.FilesystemUUID
+		NewDrive.Status.Topology = OldDrive.Status.Topology
+
+		// The name of the new drive should be its FSUUID and it should not be empty
+		if NewDrive.ObjectMeta.Name == NewDrive.Status.FSUUID && len(NewDrive.Status.FSUUID) > 0 {
```

Comment:
We can't assume that the name of the drive would necessarily be FSUUID. The older objects migrated from v1alpha and v1beta1 will not have their names as FSUUIDs